dev/OAuth/AuthRequestParams.cpp

```diff
@@ -0,0 +1,254 @@
+    }
+
+    winrt::hstring AuthRequestParams::State()
+    {
+        std::shared_lock guard{ m_mutex };
+        return m_state;
+    }
+
+    void AuthRequestParams::State(const winrt::hstring& value)
+    {
+        std::lock_guard guard{ m_mutex };
+        check_not_finalized();
+        m_state = value;
+    }
+
+    winrt::hstring AuthRequestParams::Scope()
+    {
+        std::shared_lock guard{ m_mutex };
+        return m_scope;
+    }
+
+    void AuthRequestParams::Scope(const winrt::hstring& value)
+    {
+        std::lock_guard guard{ m_mutex };
+        check_not_finalized();
+        m_scope = value;
+    }
+
+    winrt::hstring AuthRequestParams::CodeChallenge()
+    {
+        std::shared_lock guard{ m_mutex };
+        return m_codeChallenge;
+    }
+
+    void AuthRequestParams::CodeChallenge(const winrt::hstring& value)
+    {
+        std::lock_guard guard{ m_mutex };
+        check_not_finalized();
+        m_codeChallenge = value;
+    }
+
+    CodeChallengeMethodKind AuthRequestParams::CodeChallengeMethod()
+    {
+        std::shared_lock guard{ m_mutex };
+        return m_codeChallengeMethod;
+    }
+
+    void AuthRequestParams::CodeChallengeMethod(CodeChallengeMethodKind value)
+    {
+        std::lock_guard guard{ m_mutex };
+        check_not_finalized();
+        m_codeChallengeMethod = value;
+    }
+
+    IMap<winrt::hstring, winrt::hstring> AuthRequestParams::AdditionalParams()
+    {
+        std::shared_lock guard{ m_mutex };
+        return *m_additionalParams;
+    }
+
+    void AuthRequestParams::finalize()
+    {
+        std::lock_guard guard{ m_mutex };
+        if (m_finalized)
+        {
+            throw winrt::hresult_illegal_method_call(L"AuthRequestParams can only be used for a single request call");
+        }
+
+        m_finalized = true;
+        m_additionalParams->lock();
+
+        if (!m_codeChallenge.empty() && (m_codeChallengeMethod == CodeChallengeMethodKind::None))
+        {
+            throw winrt::hresult_illegal_method_call(
+                L"'CodeChallenge' cannot be set when 'CodeChallengeMethod' is set to 'None'");
+        }
+    }
+
+    void AuthRequestParams::set_state(winrt::hstring value)
+    {
+        std::lock_guard guard{ m_mutex };
+        WINRT_ASSERT(m_state.empty());
+        m_state = std::move(value);
+    }
+
+    void AuthRequestParams::set_code_challenge(winrt::hstring value)
+    {
+        std::lock_guard guard{ m_mutex };
+        WINRT_ASSERT(m_codeChallenge.empty());
+        WINRT_ASSERT(m_codeChallengeMethod != CodeChallengeMethodKind::None);
+        m_codeChallenge = std::move(value);
+    }
+
+    std::wstring AuthRequestParams::create_url(const Uri& authEndpoint)
+    {
+        std::shared_lock guard{ m_mutex };
+        WINRT_ASSERT(m_finalized);
+
+        // Per RFC 6749 section 3.1, the auth endpoint URI *MAY* contain a query string, which must be retained
+        std::wstring result{ authEndpoint.RawUri() };
```

Comment:
I'm surprised there isn't a "URI builder" that would do this for you! Looks fine as-is.